src/parser/expression.js

```diff
@@ -1080,29 +1181,43 @@ export default class ExpressionParser extends LValParser {
 
   parseIdentifier(liberal?: boolean): N.Identifier {
     const node = this.startNode();
+    const name = this.parseIdentifierName(node.start, liberal);
+    node.name = name;
+    node.loc.identifierName = name;
+    return this.finishNode(node, "Identifier");
+  }
+
+  parseIdentifierName(pos: number, liberal?: boolean): string {
     if (!liberal) {
       this.checkReservedWord(this.state.value, this.state.start, !!this.state.type.keyword, false);
     }
 
+    let name: string;
+
     if (this.match(tt.name)) {
-      node.name = this.state.value;
+      name = this.state.value;
     } else if (this.state.type.keyword) {
-      node.name = this.state.type.keyword;
+      name = this.state.type.keyword;
     } else {
-      this.unexpected();
+      throw this.unexpected();
     }
 
-    if (!liberal && node.name === "await" && this.state.inAsync) {
-      this.raise(node.start, "invalid use of await inside of an async function");
+    if (!liberal && name === "await" && this.state.inAsync) {
+      this.raise(pos, "invalid use of await inside of an async function");
     }
 
-    node.loc.identifierName = node.name;
-
     this.next();
-    return this.finishNode(node, "Identifier");
+    return name;
   }
 
   checkReservedWord(word: string, startLoc: number, checkKeywords: boolean, isBinding: boolean): void {
+    if (this.hasPlugin("typescript")) {
+      // Don't bother checking for TypeScript code.
+      // Strict mode words may be allowed as in `declare namespace N { const static: number; }`.
+      // And we have a type checker anyway, so don't bother having the parser do it.
+      return;
+    }
```

Comment:
What about a better integration with TypeScript here? Early grammar errors are great for online editor (like Babel's REPL). Babylon has that kind of logic for JavaScript.